File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        
        int[] parent = new int[enroll.length+1];
        int[] money = new int[enroll.length];
        
        HashMap<String, Integer> number = new HashMap<>();
        
        for(int i=0; i<enroll.length; i++){
            number.put(enroll[i], i+1);
        }
        
        for(int i=0; i<referral.length; i++){
            if(referral[i].equals("-")) continue;
            parent[i+1] = number.get(referral[i]);
        }
        
        for(int i=0; i<seller.length; i++){
            int num = number.get(seller[i]);
            int price = amount[i] * 100;
            while(num != 0){
                int give = price / 10;
                money[num-1] += price - give;
                num = parent[num];
                price = give;
            }
        }
        
    
        return money;
    }
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
    
    var name = new Map();
    var N = enroll.length;
    var parent = Array(N+1).fill(0);
    var result = Array(N).fill(0);
    
    for(let i=0; i<N; i++){
        name.set(enroll[i], i+1);
    }
    
    for(let i=0; i<N; i++){
        if(referral[i] == '-'){
            parent[i+1] = 0;
        }else{
            parent[i+1] = name.get(referral[i]);
        }
    }
    
    for(let i=0; i<seller.length; i++){
        var price = amount[i]*100;
        var node = name.get(seller[i]);
        while(node != 0){
            var minus = Math.floor(price/10);
            result[node-1] += price - minus;
            price = minus;
            node = parent[node];
        }
    }
    
    return result;
}